chain/ethereum/src/env.rs

```diff
@@ -0,0 +1,221 @@
+use core::ops::Deref;
+use envconfig::Envconfig;
+use graph::env::EnvVarBoolean;
+use graph::parking_lot;
+use graph::prelude::{envconfig, lazy_static, BlockNumber};
+use parking_lot::{RwLock, RwLockReadGuard};
+use std::time::Duration;
+
+lazy_static! {
+    pub static ref ENV_VARS: EnvVars = EnvVars::from_env().unwrap();
+}
+
+struct Inner {
+    env_vars: EnvVarsEthereum,
+    geth_eth_call_errors: Vec<String>,
+}
+
+impl Inner {
+    fn from_env() -> Result<Self, envconfig::Error> {
+        let env_vars = EnvVarsEthereum::init_from_env()?;
+
+        let geth_eth_call_errors = env_vars
+            .geth_eth_call_errors
+            .split(';')
+            .filter(|s| !s.is_empty())
+            .map(str::to_string)
+            .collect();
+
+        Ok(Self {
+            env_vars,
+            geth_eth_call_errors,
+        })
+    }
+}
+
+pub struct EnvVars {
+    inner: RwLock<Inner>,
+}
+
+impl EnvVars {
+    pub fn from_env() -> Result<Self, envconfig::Error> {
+        Ok(Self {
+            inner: RwLock::new(Inner::from_env()?),
+        })
+    }
+
+    /// Refreshes all internally stored environment variables to reflect the
+    /// current environment variable values.
+    ///
+    /// Only available in debug builds.
+    #[cfg(debug_assertions)]
+    pub fn refresh(&self) -> Result<(), envconfig::Error> {
```

Comment:
I believe we decided to go with setters rather than a blanket refresh. My opinions on that:

1. Setters should be added on demand, most env vars are never modified by any test.
2. We should avoid the locks, especially returning lock guards, and especially a coarse lock like this one. It means that in theory borrowing an env var value can block any other code trying to borrow any other env var.
3. Locks should be added at the env var field level, only if the env var has a setter and is not a primitive that has a corresponding atomic type.

Reply:
I think we could simplify this design a lot by just having a normal structure with public fields. If you need to test it you can pass the vars as an argument. The static reference should be kept strictly for compatibility and never really added in new code IMO. 

This need for sync is just a byproduct of the wrong design pattern. Env vars should be immutable and they should be a structure that can be built as an argument (maybe we need a build for this, maybe we use a trait so you can mock it), IMO a simple struct is much easier to understnad and use. If we implement Default for that it also helps when passing these for tests.

Reply:
Bottomline is that these envs should NOT be built to be mutated, at all IMO

Reply:
Forgot about the `refresh` thing, will update shorty :smile:  thanks for noticing.

As for locking, I agree it's not ideal, but it's a `RwLock` so while there is some overhead indeed, I'd expect the overhead to be constant and not grow with the number of readers, which won't be blocked. Sure, I can do finer-grained locking, even using atomics, it seems a bit like a premature optimization to me, that's all.

> It means that in theory borrowing an env var value can block any other code trying to borrow any other env var.

Am I missing something? It's not a mutex, so that won't happen

Reply:
@mangas if we decide *not* to allow any mutability/refreshing/setters for testing, then I'd absolutely use `pub` fields, I just thought that wasn't the team's consensus at the current time. We can reevaluate at standup today if you feel strongly about this.

Reply:
I think we didn't agree in making these arguments everywhere, which I kind of understand because it's prolly a lot of work and a big refactor for little value. That being said I would say if we need to test it we should refactor the relevant code instead of baking in an "easy way" to get around doing it properly.

Reply:
Ok, understood! I believe the only code that needs this pattern in order to be tested rn is the IPFS link resolver, so we can do that. So basically you're suggesting to modify the IPFS link resolver and immediately forgo the whole mutability thing?

Reply:
I'm a bit worried about having some areas of the codebase using the `lazy_static` and some others having env vars as extra arguments though, I feel it might be confusing.

Reply:
Yes I think we should fix whatever is immediate and remove the mutability. As for the rest I guess we could check afterwards how big of a refactor it would be to remove the `lazy_static` but at the same rate it it's probably still an improvement and from there we can just use the boyscout rule. Any new code will follow the new pattern and any old code we touch can be brought into the current pattern.

Reply:
Okay. @leoyvens how do you feel this?

Reply:
My comment on readers blocking each other was wrong because I was thinking of a mutex. If there is a simple refactor for the unit tests that would avoid the need for setters then we can just do that. I didn't understand where we would still need lazy_static (except for `EnvVars`).

In terms of the ideal ultimate solution, I'm not sure yet. Much of this code is called from multiple binaries (graph node, tests, examples, matchstick...) and it is desirable that they can all read the same env vars without re-declaring them. But for testability we want configs to be as arguments so they can be overridden.

Building on @mangas suggestion, perhaps each thing should eventually be refactored to have a `FooConfig`, and `FooConfig::default()` reads from the `EnvVars` globals. This way we have global defaults, that can set by globally defined env vars, but can still override them in tests.

Reply:
if we're all in agreement I think we should just remove the complexity, leave an empty struct and implement a Default. We can fix forward from that point 🤔

Reply:
By complexity you mean remove the getters? It does seem like if there are no setters and no locks, that they are not necessary.

Reply:
I'm not sure `EnvVars` needs to implement `Default`, as my suggestion is to have another layer of config structs, per component that needs configuration to be overridden by tests, which implements a `Default` that reads from `EnvVars`.

Reply:
Right, what I mean is EnvVars should not exist at all. As I understand it, it was just there to provide the locking and abstraction. The EthereumEnvs should have defaults as they also currently do. The packages that use these may choose to have some internal config that is generate from these. But these should be the source of truth for how default is configured, specifically at some point in the future we could generate the flag documentation from it as @tilacog suggested